Project 9 Submission

HTMLGitHub opened the pull request:
add api functionality for:
avatar image
profile section
cards

create modal for updating profile image

add button to profile image to trigger modal

HTMLGitHub committed:
add api functionality for:
avatar image
profile section
cards
create modal for updating profile image
add button to profile image to trigger modal

gennady-bars commented on src/utils/Api.js:
```diff
@@ -0,0 +1,130 @@
+    const res = await fetch(`${this.#url}/cards`, {
+      headers: this.#headers,
+    });
+    if (res.ok) {
+      return res.json();
+    }
+    return await Promise.reject(`Error: ${res.status}`);
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.


```js
_checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(this._checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the method rather than the call. 

**COULD BE IMPROVED**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

HTMLGitHub replied:
Created new function in api.js to handle the heavy lifting, and the called functions call the main function

gennady-bars commented on src/pages/index.js:
```diff
@@ -101,20 +99,72 @@ function operatePopup(popup, action="remove") {
+
+    toggleLikeStatus(data._id, isLiked);
+
     cardLikeButton.classList.toggle("card__content-like_liked");
```
You need to **toggle likes** only in a `then` block after a successful response from the server because if there is an error, the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/pages/index.js:
```diff
@@ -137,13 +187,31 @@ function addCardToList(element, method = "prepend") {
 }
 
+// Render initial cards
+function callApi() {
```
`callApi`


You get the initial app data here


`getAppData` or `getInitialData`

gennady-bars commented on src/pages/index.js:
```diff
@@ -137,13 +187,31 @@ function addCardToList(element, method = "prepend") {
 
+// Render initial cards
+function callApi() {
+  api
+    .getAPIInfo()
+    .then(([userProfile, cards ]) => {
```
Good job that you get the whole info with `Promise.all`

gennady-bars commented on src/pages/index.js:
```diff
@@ -161,22 +229,81 @@ function fillProfileForm() {
+
+  const submitButton = evt.submitter;
+
+  submitButton.textContent = "Saving...";
```
**COULD BE IMPROVED**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **resetting** and **catching errors**


```js
// define a function for changing the button text. It accepts 4 params (the 2 last are optional with default texts)
export function renderLoading(isLoading, button, buttonText='Save', loadingText='Saving...') {
  if (isLoading) {
    button.textContent = loadingText
  } else {
    button.textContent = buttonText
  }
}

// define a universal function that accepts a request function, event and a default loading text 
function handleSubmit(request, evt, loadingText = 'Saving...') {
 // You need to prevent the default action in any submit handler
  evt.preventDefault();

  // the button is always available inside `event` as `submitter`
  const submitButton = evt.submitter;
  // fix the initial button text
  const initialText = submitButton.textContent;
  // change the button text before requesting
  renderLoading(true, submitButton, initialText, loadingText);
  // call the request function to be able to use the promise chain
  request()
    .then(() => {
      // any form should be reset after a successful response 
      // evt.target is the form in any submit handler		
      evt.target.reset();
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
    .catch(console.error)      

    // and in finally we need to stop loading
    .finally(() => {
      renderLoading(false, submitButton, initialText);
    });
}

```

Here is an example of handling the profile form submit:

```js
function handleProfileFormSubmit(evt) {
  // create a request function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return editProfile(nameInput.value, jobInput.value).then((userData) => {
      userName.textContent = userData.name;
      userJob.textContent = userData.about;
    });
  }
  // here we call handleSubmit passing the request and event (if you want a different loading text then you need to pass the 3rd param)
  handleSubmit(makeRequest, evt);
}

```

So, this way you can remove a lot of code duplicating. You will not need to search buttons, pass initial button texts and so on.

`handleSubmit` and `renderLoading` should be placed in `utils.js`, because they are utility functions

gennady-bars commented on src/pages/index.js:
```diff
@@ -161,22 +229,81 @@ function fillProfileForm() {
+  api.editUserAvatar(avatarInput.value)
+    .then((user) => {
+      avatarElement.src = user.avatar;
+      evt.target.reset();
```
* Please, disable the `avatar` save button as well after submission


`disableButton` should be used here as well

HTMLGitHub replied:
got the button to disable after submission

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Matthew!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

**Parts I really liked in your project:**

* Good job that you **close** modals only in blocks `then`
* Good job that you `catch` possible errors at the end of server requests.
* Good job that you return the button text in `finally`
* Good job that you get the whole info with `Promise.all`

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:
* When I reload the page, all **black** hearts become **white** again.
* Please, disable the `avatar` save button as well after submission
* The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.
* You need to **toggle likes** only in a `then` block after a successful response from the server because if there is an error, the user will not know about that and will see the toggled heart color.
* Please, make the `confirmation` modal match the design on 320px (mobile)

There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

HTMLGitHub committed:
update delete modal
> update like status to server
> update like statas when loaded
> update api call in index.js

HTMLGitHub committed:
update callApi function name

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
